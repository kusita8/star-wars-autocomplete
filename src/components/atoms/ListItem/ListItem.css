.listItem {
  display: flex;
  align-items: center;
  height: 3rem;
  padding: 0 0.4rem;
  background: var(--yellow);
  transition: color var(--transition), background var(--transition);
  border-bottom: 1px solid black;
}

.listItem:last-of-type {
  border-bottom: none;
}

.listItem:hover {
  background: var(--red);
  color: var(--white);
}

.listItem__highlight {
  color: var(--orange);
}
