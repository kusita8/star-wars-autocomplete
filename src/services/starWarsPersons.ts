import { ApiResponse } from "../types/ApiResponse.type";
import { Person } from "../types/Person.interface";

export const getStarWarsPersons = async (name?: string): Promise<ApiResponse<Person[]>> => {
  const data = await fetch(`https://swapi.dev/api/people${name ? `?search=${name}` : ""}`);
  return data.json();
};
