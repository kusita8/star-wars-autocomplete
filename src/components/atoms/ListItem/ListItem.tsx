import { FC, useMemo } from "react";
import { Person } from "../../../types/Person.interface";
import "./ListItem.css";

interface ListItemProps {
  person: Person;
  search?: string;
}

export const ListItem: FC<ListItemProps> = ({ person, search }) => {
  const formattedSearch = search?.trim().toLocaleLowerCase();

  const linkText = useMemo(() => {
    if (!formattedSearch) return person.name;

    return person.name
      .split("")
      .map((letter) => {
        if (letter === " ") {
          return "\xa0";
        } else if (formattedSearch.includes(letter.toLocaleLowerCase())) {
          return `<span class="listItem__highlight">${letter}</span>`;
        }
        return letter;
      })
      .join("");
  }, [formattedSearch]);

  return (
    <a
      className="listItem"
      href={person.url}
      target="_blank"
      rel="noopener noreferrer"
      dangerouslySetInnerHTML={{ __html: linkText }}
    />
  );
};
