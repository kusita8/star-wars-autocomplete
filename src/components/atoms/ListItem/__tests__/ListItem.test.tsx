import { render } from "@testing-library/react";
import { getStarWarsPersons } from "../../../../services/__mocks__/starWarsPersons";
import { ListItem } from "../ListItem";

describe("ListItem", () => {
  test("it should match the snapshot", async () => {
    const { asFragment } = render(<ListItem person={(await getStarWarsPersons()).results[0]} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
