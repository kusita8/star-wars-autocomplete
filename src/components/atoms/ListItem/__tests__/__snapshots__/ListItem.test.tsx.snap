// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem it should match the snapshot 1`] = `
<DocumentFragment>
  <a
    class="listItem"
    href="https://swapi.dev/api/people/1/"
    rel="noopener noreferrer"
    target="_blank"
  >
    Luke Skywalker
  </a>
</DocumentFragment>
`;
